pennylane/ops/functions/matrix.py

```diff
@@ -118,35 +124,64 @@ def cost(theta):
         >>> qml.grad(cost)(theta)
         -0.14943813247359922
     """
+    if not isinstance(op, qml.operation.Operator):
+        if isinstance(op, (qml.tape.QuantumScript, qml.QNode)) or callable(op):

```

Comment:
I don't think you need that if you have the single dispatch?

Reply:
We do if we use it on a qnode or qfunc, because the signature of the transform only includes `QuantumTape`

Reply:
That's why I thought you need to reverse the order of register and transform

Reply:
I'll try that again just to be sure